pkg/api_builder/lib/src/client_library_generator.dart

```diff
@@ -0,0 +1,211 @@
+import 'package:code_builder/code_builder.dart' as code;
+import 'package:code_builder/code_builder.dart' show Code;
+import 'package:analyzer/dart/element/element.dart' show ClassElement;
+import 'package:analyzer/dart/element/type.dart'
+    show DartType, ParameterizedType;
+import 'package:shelf/shelf.dart' as shelf;
+import 'package:source_gen/source_gen.dart' as g;
+
+import 'shared.dart' show EndPointGenerator, Handler;
+
+code.Reference _referToType(DartType type) =>
+    code.refer(type.name, type.element.source.uri.toString());
+
+final _responseType = g.TypeChecker.fromRuntime(shelf.Response);
+
+class ClientLibraryGenerator extends EndPointGenerator {
+  @override
+  Future<String> generateForClasses(
+    Map<ClassElement, List<Handler>> classes,
+  ) async {
+    return code.Library((b) => b
+          ..directives.add(code.Directive.import(
+            'package:api_builder/_client_utils.dart',
+            show: ['ResponseException'],
+          ))
+          ..body.addAll([
+            for (final cls in classes.entries)
+              _buildClientClass(cls.key, cls.value),
+          ]))
+        .accept(
+          code.DartEmitter(code.Allocator.simplePrefixing()),
+        )
+        .toString();
+  }
+}
+
+code.Class _buildClientClass(
+  ClassElement cls,
+  List<Handler> handlers,
+) =>
+    code.Class(
+      (b) => b
+        ..docs.addAll([
+          '/// Client for invoking `${cls.name}` through the generated router.',
+          '///',
+          '/// Reponses other than 2xx causes the methods to throw',
```

Comment:
Wondering: what would happen if the response is a 300 redirect?

Reply:
I think `http.Client` follows redirects.